### PR TITLE
Linearize simulated stack

### PR DESCRIPTION
**Pull request by PalumboN**
Changes in the `reversePostOrderBlocks` algorithm to decide the order of conditional branches.

It is based on a new `stackDependency` for instructions, used to maintain the same order as the original code.

### CODE REVIEW

File: Druid/DRBasicBlock.class.st [hunk 1]
```diff
@@ -197,6 +197,24 @@ DRBasicBlock >> breakBy: instruction [
...
+{ #category : #asserting }
+DRBasicBlock >> canBeLinearizedBefore: aDRBasicBlock [
+
+	self = aDRBasicBlock ifTrue: [ ^ true ].
+
+	"Domination constraints"
+	(self isDominatedBy: aDRBasicBlock) ifTrue: [ ^ false ].
+
+	"Pops and Pushes are constraints for linearization"
+	controlFlowGraph stackInstructions do: [ :stackInstruction | "There is a stack instruction in my path ..."
```
Comment (guillep) [inline Druid/DRBasicBlock.class.st:209]: I wonder if we don't need to put this second part of the sorting in a separate (sub)class. This is only required to stage push/pop instructions of the abstract interpreter, which only happens in one of our backends.

Also, I wonder if this cannot just be rewritten a bit simpler as a topological sort with extra constraints.

File: Druid/DRBytecodeIRGenerator.class.st [hunk 2]
```diff
@@ -522,6 +522,13 @@ DRBytecodeIRGenerator >> pushFrameForCode: aDRCode receiver: aReceiver [
 DRBytecodeIRGenerator >> receiver: aDRValue [
 ]
 
+{ #category : #visiting }
+DRBytecodeIRGenerator >> trackStackInstruction: aDRStrackInstruction [ 
+
+	aDRStrackInstruction stackDependency: executionState lastStackInstruction.
+	executionState lastStackInstruction: aDRStrackInstruction.
```
Comment (guillep) [inline Druid/DRBytecodeIRGenerator.class.st:529]: :D

File: Druid/DRCogitCodeGenerator.class.st [hunk 3]
```diff
@@ -905,13 +905,6 @@ DRCogitCodeGenerator >> visitMultiplyOverflow: aDRMultiplyOverflow inBranch: aDR
 	^ aDRBranchIfCondition
 ]
 
-{ #category : #visiting }
-DRCogitCodeGenerator >> visitMultiplyOverflowJump: aDROverflowJump [ 
-	
-	self generateJump: #JumpMultiplyOverflow: target: aDROverflowJump trueBranch.
-	self generateJump: #JumpMultiplyNoOverflow: target: aDROverflowJump falseBranch.
-]
-
```
Comment (guillep) [inline Druid/DRCogitCodeGenerator.class.st:914]: Why was this removed? STRANGEEE

File: Druid/DRIRGenerator.class.st [hunk 4]
```diff
@@ -42,11 +42,13 @@ DRIRGenerator >> addFrameReturn: aDRStackFrame [
...
 { #category : #visiting }
 DRIRGenerator >> addInstruction: instruction from: aNode [
-	 
+
 	instruction origin: thisContext sender stack.
 	self currentBasicBlock addInstruction: instruction.
 	instruction originAST: aNode.
 	self pushOperand: instruction.
+	instruction isStackInstruction ifTrue: [
+		self trackStackInstruction: instruction ].
```
Comment (guillep) [inline Druid/DRIRGenerator.class.st:51]: You're lazy xD You could have put this in the push and pop only, no?

File: Druid/DRMultiplyOverflow.class.st [hunk 5]
```diff
@@ -16,16 +16,14 @@ DRMultiplyOverflow >> isMultiplyOverflow [
...
 	^ 'Multiply Overflow'
 ]
-
-{ #category : #printing }
-DRMultiplyOverflow >> reversed [
-
-	"Returns the same condition but with reversed operands: x = y --> y = x"
-1halt.
-	^ DREqualsThanComparison new
-]
```
Comment (guillep) [inline Druid/DRMultiplyOverflow.class.st:31]: dead codee :)

File: Druid/DROverflow.class.st [hunk 6]
```diff
@@ -47,14 +47,6 @@ DROverflow >> prettyPrintWithBranch: aDRBranchIfCondition [
...
 ]
 
-{ #category : #printing }
-DROverflow >> reversed [
-
-	"Returns the same condition but with reversed operands: x = y --> y = x"
-1halt.
-	^ DREqualsThanComparison new
-]
-
```
Comment (guillep) [inline Druid/DROverflow.class.st:57]: 💀

**Review by guillep — APPROVED** [review]
Thanks!

File: Druid/DRBasicBlock.class.st [hunk 7]
```diff
@@ -197,6 +197,24 @@ DRBasicBlock >> breakBy: instruction [
 		  self }
 ]
 
+{ #category : #asserting }
+DRBasicBlock >> canBeLinearizedBefore: aDRBasicBlock [
+
+	self = aDRBasicBlock ifTrue: [ ^ true ].
+
+	"Domination constraints"
+	(self isDominatedBy: aDRBasicBlock) ifTrue: [ ^ false ].
```
[the remark this answers is not in the document]
Comment (PalumboN) [reply Druid/DRBasicBlock.class.st:206]: Yes, a conditional jump with a backjump in one branch and the loop exit in the other. 
Maybe we can avoid that construction, but I remember other problems in that scenario... we can check.

Also, now the dominator tree is built in many places, should be nice to have it cached.

File: Druid/DRCogitCodeGenerator.class.st (hunk 3 above)
Comment (PalumboN) [reply Druid/DRCogitCodeGenerator.class.st:914]: Because now all the logic is in [this awful method](https://github.com/Alamvic/druid/blob/9a5e22191191cd8fef108d2b18f6976576f32d96/Druid/DRCogitCodeGenerator.class.st#L44)

File: Druid/DRIRGenerator.class.st (hunk 4 above)
Comment (PalumboN) [reply Druid/DRIRGenerator.class.st:51]: I did it first, but here is better, because some Generators on interpreting pop and push don't create the instructions (just affect the VMState or something like that).

.... and the code was uglier 😨

File: Druid/DRCogitCodeGenerator.class.st (hunk 3 above)
Comment (guillep) [reply Druid/DRCogitCodeGenerator.class.st:914]: you are weak

File: Druid/DRBasicBlock.class.st (hunk 1 above)
Comment (PalumboN) [reply Druid/DRBasicBlock.class.st:209]: https://github.com/Alamvic/druid/issues/91

File: Druid/DRBasicBlock.class.st (hunk 7 above)
[the remark this answers is not in the document]
Comment (PalumboN) [reply Druid/DRBasicBlock.class.st:206]: https://github.com/Alamvic/druid/issues/91
